File: be/src/main/java/com/sjd/luckyou/utils/SajuCalculator.java
package com.sjd.luckyou.utils;

import com.sjd.luckyou.dto.saju.SajuElement;
import com.sjd.luckyou.dto.saju.SajuInfoResponse;
import com.sjd.luckyou.dto.saju.SajuResponse;
import org.springframework.stereotype.Service;

import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.LocalTime;
import java.time.temporal.ChronoUnit;
import java.util.Arrays;
import java.util.Map;

@Service
public class SajuCalculator {

    // 천간, 지지 배열
    private static final Map<String, Map<String, String>> CHEONGAN_DATA = Map.ofEntries(
            Map.entry("갑", Map.of("ohaeng", "목", "chinese", "甲", "eumyang", "양")),
            Map.entry("을", Map.of("ohaeng", "목", "chinese", "乙", "eumyang", "음")),
            Map.entry("병", Map.of("ohaeng", "화", "chinese", "丙", "eumyang", "양")),
            Map.entry("정", Map.of("ohaeng", "화", "chinese", "丁", "eumyang", "음")),
            Map.entry("무", Map.of("ohaeng", "토", "chinese", "戊", "eumyang", "양")),
            Map.entry("기", Map.of("ohaeng", "토", "chinese", "己", "eumyang", "음")),
            Map.entry("경", Map.of("ohaeng", "금", "chinese", "庚", "eumyang", "양")),
            Map.entry("신", Map.of("ohaeng", "금", "chinese", "辛", "eumyang", "음")),
            Map.entry("임", Map.of("ohaeng", "수", "chinese", "壬", "eumyang", "양")),
            Map.entry("계", Map.of("ohaeng", "수", "chinese", "癸", "eumyang", "음"))
    );

    private static final Map<String, Map<String, String>> JIJI_DATA = Map.ofEntries(
            Map.entry("자", Map.of("animal", "쥐", "ohaeng", "수", "chinese", "子", "eumyang", "음")),
            Map.entry("축", Map.of("animal", "소", "ohaeng", "토", "chinese", "丑", "eumyang", "음")),
            Map.entry("인", Map.of("animal", "호랑이", "ohaeng", "목", "chinese", "寅", "eumyang", "양")),
            Map.entry("묘", Map.of("animal", "토끼", "ohaeng", "목", "chinese", "卯", "eumyang", "음")),
            Map.entry("진", Map.of("animal", "용", "ohaeng", "토", "chinese", "辰", "eumyang", "양")),
            Map.entry("사", Map.of("animal", "뱀", "ohaeng", "화", "chinese", "巳", "eumyang", "양")),
            Map.entry("오", Map.of("animal", "말", "ohaeng", "화", "chinese", "午", "eumyang", "음")),
            Map.entry("미", Map.of("animal", "양", "ohaeng", "토", "chinese", "未", "eumyang", "음")),
            Map.entry("신", Map.of("animal", "원숭이", "ohaeng", "금", "chinese", "申", "eumyang", "양")),
            Map.entry("유", Map.of("animal", "닭", "ohaeng", "금", "chinese", "酉", "eumyang", "음")),
            Map.entry("술", Map.of("animal", "개", "ohaeng", "토", "chinese", "戌", "eumyang", "양")),
            Map.entry("해", Map.of("animal", "돼지", "ohaeng", "수", "chinese", "亥", "eumyang", "양"))
    );
    private static final String[] CHEONGAN_KEYS = {"갑", "을", "병", "정", "무", "기", "경", "신", "임", "계"};
    private static final String[] JIJI_KEYS = {"자", "축", "인", "묘", "진", "사", "오", "미", "신", "유", "술", "해"};

    // 월 간지 계산용
    private static final String[] MONTH_BRANCHES = {"인", "묘", "진", "사", "오", "미", "신", "유", "술", "해", "자", "축"};

    // 년간 5개 그룹별(갑/기, 을/경, 병/신, 정/임, 무/계),
    // 12개월(1~12)에 해당하는 '월간' 배열
    private static final String[][] MONTH_STEMS_TABLE = {
            {"병", "정", "무", "기", "경", "신", "임", "계", "갑", "을", "병", "정"},
            {"무", "기", "경", "신", "임", "계", "갑", "을", "병", "정", "무", "기"},
            {"경", "신", "임", "계", "갑", "을", "병", "정", "무", "기", "경", "신"},
            {"임", "계", "갑", "을", "병", "정", "무", "기", "경", "신", "임", "계"},
            {"갑", "을", "병", "정", "무", "기", "경", "신", "임", "계", "갑", "을"}
    };

    // -----------------------
    // 일주 계산용 기준일: 1900-01-31 (갑자일)
    // -----------------------
    private static final LocalDate BASE_DATE = LocalDate.of(1900, 1, 31);

    // -----------------------
    // (시주) 일간에 따른 "23시 천간" 출발 인덱스
    // (갑/기=0, 을/경=2, 병/신=4, 정/임=6, 무/계=8)
    // -----------------------
    private static final int[] HOUR_STEM_START_23 = {0, 2, 4, 6, 8};


    private boolean isBeforeIpchun(int year, int month, int day) {
        LocalDate ipchunDate = LocalDate.of(year, 2, 4);
        LocalDate birthDate = LocalDate.of(year, month, day);
        return birthDate.isBefore(ipchunDate);
    }

    // (1) 연주 계산
    private String getGanzhiYear(int year, int month, int day) {
        boolean beforeIpchun = isBeforeIpchun(year, month, day);
        int adjustedYear = beforeIpchun ? year - 1 : year;

        int baseYear = 1924; // 1924년 = 갑자년
        int gap = adjustedYear - baseYear;
        int stemIndex = ((gap % 10) + 10) % 10; // 천간
        int branchIndex = ((gap % 12) + 12) % 12; // 지지

        return CHEONGAN_KEYS[stemIndex] + JIJI_KEYS[branchIndex];
    }

    // (2) 월주 계산
    private String getGanzhiMonth(int year, int month, int day) {
        boolean beforeIpchun = isBeforeIpchun(year, month, day);
        int usedMonth = month - 1;
        int adjustedYear = beforeIpchun ? year - 1 : year;
        if (usedMonth < 0) {
            usedMonth += 12;
            adjustedYear -= 1;
        }

        int gap = adjustedYear - 1984;
        int stemIndex = ((gap % 10) + 10) % 10;
        String yearStem = CHEONGAN_KEYS[stemIndex];

        int stemGroup = switch (yearStem) {
            case "갑", "기" -> 0;
            case "을", "경" -> 1;
            case "병", "신" -> 2;
            case "정", "임" -> 3;
            case "무", "계" -> 4;
            default -> throw new IllegalArgumentException("Invalid yearStem = " + yearStem);
        };

        return MONTH_STEMS_TABLE[stemGroup][usedMonth - 1] + MONTH_BRANCHES[usedMonth - 1];
    }

    // (3) 일주 계산
    private String getGanzhiDay(LocalDateTime dateTime) {
        long days = ChronoUnit.DAYS.between(BASE_DATE, dateTime.toLocalDate());
        int dayIndex = (int) ((days + 40) % 60);
        if (dayIndex < 0) dayIndex += 60;
        return CHEONGAN_KEYS[dayIndex % 10] + JIJI_KEYS[dayIndex % 12];
    }

    // (4) 시주 계산
    private String getGanzhiHour(String dayStem, int hour) {
        int hourBranchIndex = (hour == 23 ? 0 : (hour + 1) / 2) % 12;
        String hourBranch = JIJI_KEYS[hourBranchIndex];
        int dayStemIndex = Arrays.asList(CHEONGAN_KEYS).indexOf(dayStem);
        if (dayStemIndex == -1) {
            System.err.println("일간(천간) 찾기 실패: " + dayStem + ", 기본값 '갑' 사용");
            return CHEONGAN_KEYS[0] + hourBranch;
        }
        int stemAt23 = HOUR_STEM_START_23[dayStemIndex % 5];
        int hourStemIndex = (stemAt23 + hourBranchIndex) % 10;
        return CHEONGAN_KEYS[hourStemIndex] + hourBranch;
    }

    private SajuElement createSajuElement(String gan, String ji) {
        return SajuElement.builder()
                .gan(gan)
                .ji(ji)
                .animal(ji.equals("-") ? "-" : JIJI_DATA.get(ji).get("animal"))
                .ganOhaeng(gan.equals("-") ? "-" : CHEONGAN_DATA.get(gan).get("ohaeng"))
                .ganChinese(gan.equals("-") ? "-" : CHEONGAN_DATA.get(gan).get("chinese"))
                .jiOhaeng(ji.equals("-") ? "-" : JIJI_DATA.get(ji).get("ohaeng"))
                .jiChinese(ji.equals("-") ? "-" : JIJI_DATA.get(ji).get("chinese"))
                .build();
    }

    public SajuResponse calculateSaju(SajuInfoResponse request) {
        if (request.getBirthDate() == null) {
            throw new IllegalArgumentException("birthDate는 필수입니다.");
        }

        LocalDateTime birthDateTime = LocalDateTime.of(request.getBirthDate(),
                request.getBirthTime() != null ? request.getBirthTime() : LocalTime.of(0, 0));
        // 한국 기준 30분 보정
        birthDateTime = birthDateTime.minusMinutes(30);

        int year = birthDateTime.getYear();
        int month = birthDateTime.getMonthValue();
        int day = birthDateTime.getDayOfMonth();
        int hour = birthDateTime.getHour();

        String yearPillar = getGanzhiYear(year, month, day);
        String monthPillar = getGanzhiMonth(year, month, day);
        String dayPillar = getGanzhiDay(birthDateTime);
        String timePillar = request.getBirthTime() != null ? getGanzhiHour(dayPillar.substring(0, 1), hour) : "-";

        return SajuResponse.builder()
                .id(request.getId())
                .year(createSajuElement(yearPillar.substring(0, 1), yearPillar.substring(1)))
                .month(createSajuElement(monthPillar.substring(0, 1), monthPillar.substring(1)))
                .day(createSajuElement(dayPillar.substring(0, 1), dayPillar.substring(1)))
                .time(createSajuElement(timePillar.equals("-") ? "-" : timePillar.substring(0, 1), timePillar.equals("-") ? "-" : timePillar.substring(1)))
                .build();
    }
}

File: fe/utils/sajuCalculator.ts
import { SajuElement, SajuInfo } from '@/types/Saju';
import { daysBetween } from './timer';

const CHEONGAN_DATA: Record<string, { ohaeng: string; chinese: string; eumyang: string }> = {
  갑: { ohaeng: '목', chinese: '甲', eumyang: '양' },
  을: { ohaeng: '목', chinese: '乙', eumyang: '음' },
  병: { ohaeng: '화', chinese: '丙', eumyang: '양' },
  정: { ohaeng: '화', chinese: '丁', eumyang: '음' },
  무: { ohaeng: '토', chinese: '戊', eumyang: '양' },
  기: { ohaeng: '토', chinese: '己', eumyang: '음' },
  경: { ohaeng: '금', chinese: '庚', eumyang: '양' },
  신: { ohaeng: '금', chinese: '辛', eumyang: '음' },
  임: { ohaeng: '수', chinese: '壬', eumyang: '양' },
  계: { ohaeng: '수', chinese: '癸', eumyang: '음' },
};

// 천간(10개): 갑, 을, 병, 정, 무, 기, 경, 신, 임, 계
const CHEONGAN_KEYS = Object.keys(CHEONGAN_DATA);

const JIJI_DATA: Record<string, { animal: string; ohaeng: string; chinese: string; eumyang: string }> = {
  자: { animal: '쥐', ohaeng: '수', chinese: '子', eumyang: '음' },
  축: { animal: '소', ohaeng: '토', chinese: '丑', eumyang: '음' },
  인: { animal: '호랑이', ohaeng: '목', chinese: '寅', eumyang: '양' },
  묘: { animal: '토끼', ohaeng: '목', chinese: '卯', eumyang: '음' },
  진: { animal: '용', ohaeng: '토', chinese: '辰', eumyang: '양' },
  사: { animal: '뱀', ohaeng: '화', chinese: '巳', eumyang: '양' },
  오: { animal: '말', ohaeng: '화', chinese: '午', eumyang: '음' },
  미: { animal: '양', ohaeng: '토', chinese: '未', eumyang: '음' },
  신: { animal: '원숭이', ohaeng: '금', chinese: '申', eumyang: '양' },
  유: { animal: '닭', ohaeng: '금', chinese: '酉', eumyang: '음' },
  술: { animal: '개', ohaeng: '토', chinese: '戌', eumyang: '양' },
  해: { animal: '돼지', ohaeng: '수', chinese: '亥', eumyang: '양' },
};
// 지지(12개): 자, 축, 인, 묘, 진, 사, 오, 미, 신, 유, 술, 해
const JIJI_KEYS = Object.keys(JIJI_DATA);

// 월 간지 계산용
const MONTH_BRANCHES = ['인', '묘', '진', '사', '오', '미', '신', '유', '술', '해', '자', '축'];
const MONTH_STEMS_TABLE = [
  ['병', '정', '무', '기', '경', '신', '임', '계', '갑', '을', '병', '정'], // 갑, 기
  ['무', '기', '경', '신', '임', '계', '갑', '을', '병', '정', '무', '기'], // 을, 경
  ['경', '신', '임', '계', '갑', '을', '병', '정', '무', '기', '경', '신'], // 병, 신
  ['임', '계', '갑', '을', '병', '정', '무', '기', '경', '신', '임', '계'], // 정, 임
  ['갑', '을', '병', '정', '무', '기', '경', '신', '임', '계', '갑', '을'], // 무, 계
];

// 일주 계산용 기준일: 1900-01-31 (갑자일)
const BASE_DATE = new Date(1900, 0, 31);
// 시주 계산용: 일간에 따른 "23시 천간" 출발 인덱스
const HOUR_STEM_START_23 = [0, 2, 4, 6, 8];

// 연주 계산 (Year Pillar)
function getGanzhiYear(year: number, month: number, day: number): string {
  const beforeIpchun = isBeforeIpchun(year, month, day);
  const adjustedYear = beforeIpchun ? year - 1 : year;

  const baseYear = 1924; // 1924 = 갑자년
  const gap = adjustedYear - baseYear;
  const stemIndex = ((gap % 10) + 10) % 10; // 천간
  const branchIndex = ((gap % 12) + 12) % 12; // 지지

  return CHEONGAN_KEYS[stemIndex] + JIJI_KEYS[branchIndex];
}

function isBeforeIpchun(year: number, month: number, day: number): boolean {
  // TODO: 실제 입춘 날짜는 해마다 다를 수 있음. 향후 정확한 데이터로 대체 가능.
  const ipchunDate = new Date(year, 1, 4); // Feb 4 (입춘 근사치)
  const birthDate = new Date(year, month - 1, day);
  return birthDate < ipchunDate;
}

// 월주 계산 (Month Pillar)
function getGanzhiMonth(year: number, month: number, day: number): string {
  const beforeIpchun = isBeforeIpchun(year, month, day);
  let usedMonth = month - 1;
  let adjustedYear = beforeIpchun ? year - 1 : year;

  if (usedMonth < 1) {
    usedMonth += 12;
    adjustedYear -= 1;
  }

  const gap = adjustedYear - 1984;
  const stemIndex = ((gap % 10) + 10) % 10;
  const yearStem = CHEONGAN_KEYS[stemIndex];

  const stemGroup = (() => {
    switch (yearStem) {
      case '갑':
      case '기':
        return 0;
      case '을':
      case '경':
        return 1;
      case '병':
      case '신':
        return 2;
      case '정':
      case '임':
        return 3;
      case '무':
      case '계':
        return 4;
      default:
        throw new Error(`Invalid yearStem = ${yearStem}`);
    }
  })();

  const monthStem = MONTH_STEMS_TABLE[stemGroup][usedMonth - 1];
  const monthBranch = MONTH_BRANCHES[usedMonth - 1];

  return monthStem + monthBranch;
}
// 일주 계산 (Day Pillar)
function getGanzhiDay(dateTime: Date): string {
  const daysBetweenBase = daysBetween(BASE_DATE, dateTime);
  let dayIndex = (daysBetweenBase + 40) % 60;
  if (dayIndex < 0) {
    dayIndex += 60;
  }

  const stemIndex = dayIndex % 10;
  const branchIndex = dayIndex % 12;

  return CHEONGAN_KEYS[stemIndex] + JIJI_KEYS[branchIndex];
}

// 시주 계산 (Hour Pillar)
function getGanzhiHour(dayStem: string, hour: number): string {
  const hourBranchIndex = hour === 23 ? 0 : Math.floor((hour + 1) / 2) % 12;
  const hourBranch = JIJI_KEYS[hourBranchIndex];

  const dayStemIndex = CHEONGAN_KEYS.indexOf(dayStem);
  if (dayStemIndex === -1) {
    console.error(`일간(천간) 찾기 실패: ${dayStem}, 기본값 '갑' 사용`);
    return CHEONGAN_KEYS[0] + hourBranch; // 기본값으로 '갑' 사용
  }

  const group5 = dayStemIndex % 5;
  const stemAt23 = HOUR_STEM_START_23[group5];
  const hourStemIndex = (stemAt23 + hourBranchIndex) % 10;

  return CHEONGAN_KEYS[hourStemIndex] + hourBranch;
}

/**
 * 사주 계산 함수
 */
export function calculateSaju(info: SajuInfo) {
  const { birthDate, birthTime } = info;
  const [year, month, day] = birthDate.split('-').map(Number);
  const [hour, minute] = birthTime ? birthTime.split(':').map(Number) : [0, 0];

  const birthDateTime = new Date(year, month - 1, day, hour, minute);
  const adjust30Min = false; // 한국 사주에서 시간 보정 필요 시 true로 설정
  if (adjust30Min) {
    birthDateTime.setMinutes(birthDateTime.getMinutes() - 30); // 30분 조정 (선택적)
  }

  const adjustYear = birthDateTime.getFullYear();
  const adjustMonth = birthDateTime.getMonth() + 1;
  const adjustDay = birthDateTime.getDate();
  const adjustHour = birthDateTime.getHours();

  const yearPillar = getGanzhiYear(adjustYear, adjustMonth, adjustDay);
  const monthPillar = getGanzhiMonth(adjustYear, adjustMonth, adjustDay);
  const dayPillar = getGanzhiDay(birthDateTime);
  const timePillar = birthTime ? getGanzhiHour(dayPillar.substring(0, 1), adjustHour) : '-';

  const yearGan = yearPillar[0],
    yearJi = yearPillar[1];
  const monthGan = monthPillar[0],
    monthJi = monthPillar[1];
  const dayGan = dayPillar[0],
    dayJi = dayPillar[1];
  const timeGan = timePillar !== '-' ? timePillar[0] : '-',
    timeJi = timePillar !== '-' ? timePillar[1] : '-';

  return {
    year: createSajuElement(yearGan, yearJi),
    month: createSajuElement(monthGan, monthJi),
    day: createSajuElement(dayGan, dayJi),
    time: createSajuElement(timeGan, timeJi),
  };
}

// SajuElement 생성 헬퍼 함수
function createSajuElement(gan: string, ji: string): SajuElement {
  return {
    gan,
    ji,
    animal: ji !== '-' ? JIJI_DATA[ji].animal : '-',
    ganOhaeng: gan !== '-' ? CHEONGAN_DATA[gan].ohaeng : '-',
    ganChinese: gan !== '-' ? CHEONGAN_DATA[gan].chinese : '-',
    jiOhaeng: ji !== '-' ? JIJI_DATA[ji].ohaeng : '-',
    jiChinese: ji !== '-' ? JIJI_DATA[ji].chinese : '-',
  };
}
